usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+    memset(fname, 0, PATH_MAX);
+    snprintf(fname, PATH_MAX, "%s_PKCSTOK_MIGRATE_BACKUP", data_store);
+    if (stat(fname, &statbuf) != 0) {
+        return CK_FALSE;
+    }
+
+    if (!S_ISDIR(statbuf.st_mode)) {
+        TRACE_ERROR("%s already exists but is not a directory.\n", fname);
+        return CK_TRUE;
+    }
+
+    return CK_TRUE;
+}
+
+/**
+ * Copy a file given by name from a src folder to a dst folder.
+ */
+void file_copy(char *dst, const char *src, const char *name)
+{
+    char dst_file[PATH_MAX], src_file[PATH_MAX], buf[4096];
+    FILE *fp_r, *fp_w;
+
+    snprintf(dst_file, PATH_MAX, "%s/%s", dst, name);
+    snprintf(src_file, PATH_MAX, "%s/%s", src, name);
+
+    fp_r = fopen(src_file, "r");
+    fp_w = fopen(dst_file, "w");
+
+    while (!feof(fp_r)) {
+        size_t bytes = fread(buf, 1, sizeof(buf), fp_r);
+        if (bytes)
+            fwrite(buf, 1, bytes, fp_w);
+    }
+
+    fclose(fp_r);
+    fclose(fp_w);
+}
+
+/**
+ * Copy the given src folder to the given dst folder including all
+ * subdirectories and files.
+ */
+CK_RV folder_copy(char *dst, const char *src)
+{
+    char d[PATH_MAX], s[PATH_MAX];
+    struct dirent *entry;
+    CK_RV ret;
+    DIR *dir;
+
+    /* Open src */
+    dir = opendir(src);
+    if (dir == NULL) {
+        TRACE_ERROR("Cannot open %s\n", src);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /* Create dst */
+    if (mkdir(dst, 0) != 0) {
+        TRACE_ERROR("Cannot create %s\n", dst);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /* Copy folder recursively, skip the "." and ".." entries */
+    while ((entry = readdir(dir)) != NULL) {
+        if (entry->d_type == DT_DIR) {
+            if (strncmp(entry->d_name, ".", 1) != 0) {
+                snprintf(d, PATH_MAX, "%s/%s", dst, entry->d_name);
+                snprintf(s, PATH_MAX, "%s/%s", src, entry->d_name);
+                ret = folder_copy(d, s);
+                if (ret != CKR_OK)
+                    break;
+            }
+        } else {
+            file_copy(dst, src, entry->d_name);
+        }
+    }
+
+    closedir(dir);
+
+    return CKR_OK;
+}
+
+/**
+ * Backups the given data_store to data_store_PKCSTOK_MIGRATE_BACKUP.
+ * All folders and files are recursively created and copied.
+ */
+CK_RV backup_repository(const char *data_store)
+{
+    char dst[PATH_MAX];
+    struct stat statbuf;
+
+    /* Check if data_store_PKCSTOK_MIGRATE_BACKUP exists */
+    memset(dst, 0, PATH_MAX);
+    snprintf(dst, PATH_MAX, "%s_PKCSTOK_MIGRATE_BACKUP", data_store);
+    if (stat(dst, &statbuf) == 0 && S_ISDIR(statbuf.st_mode)) {
+        TRACE_INFO("Repository backup already exists, using this one.\n");
+        return CKR_OK;
+    }
+
+    TRACE_INFO("Creating backup repository ...\n");
+    return folder_copy(dst, data_store);
+}
+
+/**
+ * Checks if the pkcsslotd is running.
+ */
+CK_BBOOL pkcsslotd_running(void)
+{
+    DIR* dir;
+    struct dirent* ent;
+    char* endptr;
+    char buf[512];
```

Comment:
MAX_PATH ?

Reply:
buf shouldn't be that long, but ok, why not.

Reply:
Well, I guess the cmdline could contain an executable with its full path, so it could be at least MAX_PATH long, if not longer (plus arguments).

Reply:
There might also be another problem. Wouldn't you also find a process that was started like 'vi pkcsslotd' ? The strstr just looks for 'pkcsslotd' anywhere in the cmdline.....